### PR TITLE
Introduce a function that awaits startup with a timeout

### PR DESCRIPTION
**Pull request by michaelklishin**
## Proposed Changes

This extends `rabbit:await_startup/{0,1,2}` to be suitable for a new CLI command. It would
be an alternative to `rabbitmqctl wait [pid file]` that doesn't require a pid file path (similarly to
how `rabbitmqctl shutdown` works compared to `rabbitmqctl stop [pid file]`.

## Types of Changes

- [ ] Bugfix (non-breaking change which fixes issue #NNNN)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, appearance)

## Checklist

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in related repositories

## Further Comments

One way to test the timeout path is to temporarily modify `rabbit:broker_start/0`:

``` erlang
broker_start() ->
    Plugins = rabbit_plugins:setup(),
    ToBeLoaded = Plugins ++ ?APPS,
    start_apps(ToBeLoaded),
    maybe_sd_notify(),
    ok = rabbit_lager:broker_is_started(),
    timer:start(90000),
    ok = log_broker_started(rabbit_plugins:strictly_plugins(rabbit_plugins:active())).
```

### CODE REVIEW

**Comment by lukebakken** [comment]
Windows users will appreciate this new command since a pid file isn't created.

File: src/rabbit.erl [hunk 1]
```diff
@@ -240,6 +240,7 @@
 -define(BOOT_START_TIMEOUT,     300).
 %% 12 hours
 -define(BOOT_FINISH_TIMEOUT,    43200).
+%% 100ms
```
Comment (gerhard) [inline src/rabbit.erl]: If that is true, and `Iterations = div(BOOT_START_TIMEOUT, BOOT_STATUS_SAMPLING_INTERVAL)` in `do_wait_for_boot_to_Start/2`, then I read `BOOT_START_TIMEOUT` & `BOOT_START_TIMEOUT` to be in milliseconds, not seconds.

Comment (lukebakken) [reply src/rabbit.erl]: The names / comments could be clearer, for sure.

File: src/rabbit.erl [hunk 2]
```diff
@@ -236,6 +236,12 @@
 
 -define(ASYNC_THREADS_WARNING_THRESHOLD, 8).
 
+%% 5 minutes
+-define(BOOT_START_TIMEOUT,     300).
+%% 12 hours
+-define(BOOT_FINISH_TIMEOUT,    43200).
```
[the remark this answers is not in the document]
Comment (michaelklishin) [reply src/rabbit.erl]: We discussed this: this is a fallback timeout and CLI tools default to 5 minutes, decided to leave it as is.

File: src/rabbit.erl (hunk 1 above)
Comment (michaelklishin) [reply src/rabbit.erl]: Corrected.

File: src/rabbit.erl [hunk 3]
```diff
@@ -236,6 +236,12 @@
...
+-define(BOOT_STATUS_SAMPLING_INTERVAL, 100).
```
[the remark this answers is not in the document]
Comment (michaelklishin) [reply src/rabbit.erl]: It's 100 ms.

**Comment by michaelklishin** [comment]
Backported to `v3.7.x`.

**Comment by olamartensson** [comment]
The added CLI function `await_startup` is much appreciated on the Windows side. 👍

However, I noted that the online [man page](https://www.rabbitmq.com/rabbitmqctl.8.html) for `rabbitmqctl` is lacking a reference to this new function.

**Comment by michaelklishin** [comment]
@olamartensson we are working on updating man pages for `3.8.0`, see recent changes under the `docs` directory.

**Comment by olamartensson** [comment]
@michaelklishin: That sounds great. :) I did a search for the `await_startup` function among the files below `docs` but without any results. Perhaps I was just not on the correct branch. Anyway, now I know it will be added. Thanks.

**Comment by michaelklishin** [comment]
@olamartensson it's there in https://github.com/rabbitmq/rabbitmq-server/pull/2123/files (make sure to explicitly load the `rabbitmqctl.8` file as GitHub won't load large files in diffs by default)

**Pull request merged**